addons/account/migrations/9.0.1.1/post-migration.py

```diff
@@ -303,23 +402,196 @@ def account_partial_reconcile(env):
+        ON Q5.reconcile_id = amr.id
+        WHERE debit_move_id > 0 AND credit_move_id > 0
+    """)
+
+    # We want to exclude the moves that were included in the step above from
+    # the next reconciliation step.
+    cr.execute("""
+        WITH Q1 AS (
+            SELECT reconcile_id, sum(debit-credit) as balance,
+            count(id) as num_moves, count(currency_id) as num_currencies
+            FROM account_move_line
+            WHERE reconcile_id IS NOT NULL
+            GROUP BY reconcile_id, currency_id
+        ),
+        Q2 AS (
+            SELECT reconcile_id
+            FROM Q1
+            WHERE balance = 0.0
+            AND num_moves = 2
+            AND num_currencies = 0
+        )
+        SELECT aml.id
+        FROM account_move_line AS aml
+        INNER JOIN Q2
+        ON Q2.reconcile_id = aml.reconcile_id
+    """)
+
+    move_line_ids_reconciled = [move_line_id for move_line_id,
+                                in cr.fetchall()]
+
+    move_line_map = {}
+    cr.execute("SELECT reconcile_id, id "
+               "FROM account_move_line "
+               "WHERE reconcile_id IS NOT NULL "
+               "AND id NOT IN %s" % (tuple(move_line_ids_reconciled), ))
+    rec_l = {}
+    for rec_id, move_line_id in cr.fetchall():
+        move_line_map.setdefault(rec_id, []).append(move_line_id)
+        rec_l[rec_id] = True
+    num_recs = len(rec_l.keys())
+    i = 1
+    for _rec_id, move_line_ids in move_line_map.iteritems():
+        msg = 'Reconciliation step 2 (%s of %s). ' \
+              'Resolving account.move.reconcile %s.' % \
+              (i, num_recs, _rec_id)
+        openupgrade.message(cr, 'account', 'account_partial_reconcile',
+                            'id', msg)
+        move_lines = env['account.move.line'].browse(move_line_ids)
+        amount_residual_d = {}
+        for move_line in move_lines:
+            amount = abs(move_line.debit - move_line.credit)
+            amount_residual_currency = abs(move_line.amount_currency) or 0.0
+            sign = 1 if (move_line.debit - move_line.credit) > 0 else -1
+            amount_residual = move_line.company_id.currency_id.round(
+                amount * sign)
+            amount_residual_d[move_line.id] = {}
+            amount_residual_d[move_line.id]['amount_residual'] = \
+                amount_residual
+            amount_residual_d[move_line.id]['amount_residual_currency'] = \
+                move_line.currency_id and move_line.currency_id.round(
+                    amount_residual_currency * sign) or 0.0
+        auto_reconcile_lines(env, move_lines, amount_residual_d)
+        i += 1
+        move_line_ids_reconciled += move_line_ids
+
+    # The previous move lines must be flagged as reconciled. The residual
+    # amount is 0.
+    move_lines = env['account.move.line'].with_context(recompute=False).browse(
+        move_line_ids_reconciled)
+    for move_line in move_lines:
+        move_line.reconciled = True
```

Comment:
no, I'm on vacation now. @NL66278's branch finishes in 10 minutes for this database, that's surprisingly fast. But I didn't yet make it to check if the results are correct.

As for the problem mentioned here: Also no time to debug throug this, but the problematic move looks like this:

```sql
select * from account_move_line join account_move on move_id=account_move.id where account_move.id=16916;
```

```
  id   | create_uid |        create_date         |         write_date         | write_uid | statement_id | journal_id | currency_id | date_maturity | partner_id | reconcile_partial_id | blocked | analytic_account_id | credit | centralisation | company_id | tax_code_id | state | debit |       ref       | account_id | period_id | date_created |    date    | move_id |    name    | reconcile_id | tax_amount | product_id | account_tax_id | product_uom_id | amount_currency | quantity | last_rec_date |  id   | create_uid |        create_date         |         write_date         | write_uid |     name     | state |       ref       | company_id | journal_id | period_id | narration |    date    | partner_id | to_check | balance 
-------+------------+----------------------------+----------------------------+-----------+--------------+------------+-------------+---------------+------------+----------------------+---------+---------------------+--------+----------------+------------+-------------+-------+-------+-----------------+------------+-----------+--------------+------------+---------+------------+--------------+------------+------------+----------------+----------------+-----------------+----------+---------------+-------+------------+----------------------------+----------------------------+-----------+--------------+-------+-----------------+------------+------------+-----------+-----------+------------+------------+----------+---------
 58233 |         12 | 2012-09-19 09:40:17.634682 | 2014-07-21 08:12:12.466196 |        12 |              |          3 |             | 2012-09-30    |        711 |                      | f       |                     |  70.88 | normal         |          1 |             | draft |  0.00 | XXXXXXXXXXXXXXX |        175 |        20 | 2012-09-19   | 2012-08-23 |   16916 | XXXXXXXXXX |          976 |       0.00 |            |                |                |            0.00 |          | 2012-08-25    | 16916 |         12 | 2012-09-19 09:40:17.634682 | 2014-07-21 08:12:12.466196 |        12 | EXJ/2012/522 | draft | XXXXXXXXXXXXXXX |          1 |          3 |        20 |           | 2012-08-23 |        711 | f        |        
(1 row)
```

So you see it's actually unbalanced, but judging from the above, those should be treated differently anyways, right? (on the other hand, it's late...)

And as far as the comparison goes, it's of course apples and oranges as long as I didn't check validity on either yet, and for my test database, this code doesn't finish in the first place